infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/CorfuReplicationManager.java

```
@@ -0,0 +1,300 @@
+
+    private long totalNumEntriesToSend;
+
+    private final LogReplicationMetadataManager metadataManager;
+
+    private final String pluginFilePath;
+
+    /**
+     * Constructor
+     *
+     * @param topologyDescriptor description of active and standby cluster' of a given topology
+     * @param logReplicationConfig log replication configuration
+     */
+    public CorfuReplicationManager(TopologyDescriptor topologyDescriptor,
+                            LogReplicationConfig logReplicationConfig,
+                            NodeDescriptor localNodeDescriptor,
+                            LogReplicationMetadataManager metadataManager,
+                            String pluginFilePath, CorfuRuntime corfuRuntime) {
+        this.topologyDescriptor = topologyDescriptor;
+        this.logReplicationConfig = logReplicationConfig;
+        this.metadataManager = metadataManager;
+        this.pluginFilePath = pluginFilePath;
+        this.corfuRuntime = corfuRuntime;
+
+        this.localNodeDescriptor = localNodeDescriptor;
+        this.prepareSiteRoleChangeStreamTail = Address.NON_ADDRESS;
+        this.totalNumEntriesToSend = 0;
+    }
+
+    /**
+     * Start Log Replication Manager, this will initiate a runtime against
+     * each standby cluster, to further start log replication.
+     */
+    public void start() {
+        for (ClusterDescriptor remoteCluster : topologyDescriptor.getStandbyClusters().values()) {
+            try {
+                startLogReplicationRuntime(remoteCluster);
+            } catch (Exception e) {
+                log.error("Failed to start log replication runtime for remote cluster {}", remoteCluster.getClusterId());
+
+                // Remove cluster from the list of standby's, as the cluster discovery process will receive
+                // change notification when this site becomes stable/available again.
+                topologyDescriptor.getStandbyClusters().remove(remoteCluster.getClusterId());
+            }
```

It looks like that this catch block is in a dead code path. Because startLogReplicationRuntime will catch all exceptions and input an event to FSM.

Remove elements in an iteration over values also dangerous.

> If the map is modified while an iteration over the collection is in progress (except through the iterator's own remove operation), the results of the iteration are undefined.